src/openqaoa-core/derivatives/derivative_functions.py

```diff
@@ -1,202 +1,254 @@
-#   Copyright 2022 Entropica Labs
-#
-#   Licensed under the Apache License, Version 2.0 (the "License");
-#   you may not use this file except in compliance with the License.
-#   You may obtain a copy of the License at
-#
-#       http://www.apache.org/licenses/LICENSE-2.0
-#
-#   Unless required by applicable law or agreed to in writing, software
-#   distributed under the License is distributed on an "AS IS" BASIS,
-#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-#   See the License for the specific language governing permissions and
-#   limitations under the License.
-
 from __future__ import annotations
```

Comment:
All the derivative functions are only used by the optimizers, aren't they? Why don't we put this folder 'derivatives' inside the 'optimizers' folder?

Reply:
Makes sense. I can't think of a reason why derivatives would be better placed in its current heriarchy other than the possibility of needing to call it from out openqaoa, and making that call easier to do so.

@Q-lds @vishal-ph Quick comments here would help. Thanks.

Reply:
We wont move the folder

Reply:
what was the reason at the end? to me it is not intuitive to have it at the same level as algorithms, backends, optimizers, problems, components